### PR TITLE
add etherscan plugin

### PR DESCRIPTION
**Pull request by yann300**

### CODE REVIEW

File: apps/etherscan/src/app/AppContext.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import React from "react"
+import { PluginClient } from "@remixproject/plugin"
+
+import { Receipt, ThemeType } from "./types"
+
+export const AppContext = React.createContext({
+  apiKey: "",
+  setAPIKey: (value: string) => {
+    console.log("Set API Key from Context")
```
Comment (Aniket-Engg) [inline apps/etherscan/src/app/AppContext.tsx:9]: are consoles in this file necessary?

File: apps/etherscan/src/app/app.tsx [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  const [contracts, setContracts] = useState([] as string[])
+  const [themeType, setThemeType] = useState("dark" as ThemeType)
+
+  const clientInstanceRef = useRef(clientInstance)
+  clientInstanceRef.current = clientInstance
+  const contractsRef = useRef(contracts)
+  contractsRef.current = contracts
+
+  useEffect(() => {
+    console.log("Remix Etherscan loading...")
```
Comment (Aniket-Engg) [inline apps/etherscan/src/app/app.tsx:46]: here too

File: apps/etherscan/src/app/app.tsx [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    const client = new PluginClient()
+    createClient(client)
+    const loadClient = async () => {
+      await client.onload()
+      setClientInstance(client)
+      console.log("Remix Etherscan Plugin has been loaded")
```
Comment (Aniket-Engg) [inline apps/etherscan/src/app/app.tsx:52]: this log

File: apps/etherscan/src/app/hooks/useLocalStorage.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  // Pass initial state function to useState so logic is only executed once
+  const [storedValue, setStoredValue] = useState(() => {
+    try {
+      // Get from local storage by key
+      const item = window.localStorage.getItem(key)
+      // Parse stored json or if none return initialValue
+      return item ? JSON.parse(item) : initialValue
+    } catch (error) {
+      // If error also return initialValue
+      console.log(error)
```
Comment (Aniket-Engg) [inline apps/etherscan/src/app/hooks/useLocalStorage.tsx:14]: better to use console.error

File: apps/etherscan/src/app/hooks/useLocalStorage.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      // Allow value to be a function so we have same API as useState
+      const valueToStore =
+        value instanceof Function ? value(storedValue) : value
+      // Save state
+      setStoredValue(valueToStore)
+      // Save to local storage
+      window.localStorage.setItem(key, JSON.stringify(valueToStore))
+    } catch (error) {
+      // A more advanced implementation would handle the error case
+      console.log(error)
```
Comment (Aniket-Engg) [inline apps/etherscan/src/app/hooks/useLocalStorage.tsx:32]: here too

File: apps/etherscan/src/app/utils/utilities.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+export const getNetworkName = async (client: RemixClient) => {
+  const network = await client.call("network", "detectNetwork")
+  if (!network) {
+    throw new Error("no known network to verify against")
+  }
+  const name = network.name!.toLowerCase()
+  // TODO : remove that when https://github.com/ethereum/remix-ide/issues/2017 is fixe
+  return name === "görli" ? "goerli" : name
```
Comment (Aniket-Engg) [inline apps/etherscan/src/app/utils/utilities.ts:18]: is this still needed?

File: apps/etherscan/src/app/views/ErrorView.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        width="250"
+        src="https://res.cloudinary.com/key-solutions/image/upload/v1580400635/solid/error-png.png"
+        alt="Error page"
+      />
+      <h5>Sorry, something unexpected happened. </h5>
+      <h5>
+        Please raise an issue:{" "}
+        <a
+          style={{ color: "red" }}
+          href="https://github.com/machinalabs/remix-etherscan/issues"
```
Comment (Aniket-Engg) [inline apps/etherscan/src/app/views/ErrorView.tsx:24]: May be this link should be updated

File: apps/remix-ide-e2e/src/tests/etherscan_api.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  'Should load etherscan plugin #group1': function (browser: NightwatchBrowser) {
+    browser.clickLaunchIcon('pluginManager')
+      .scrollAndClick('[data-id="pluginManagerComponentActivateButtonetherscan"]')
+      .clickLaunchIcon('etherscan')
+      .pause(5000)
+      // @ts-ignore
+      .frame(0)
+      .waitForElementNotVisible('input[name="apiKey"]')
+  }  
+}
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/etherscan_api.ts:24]: We should try to add more tests

**Review by bunsenstraat — APPROVED** [review]

**Comment by Aldekein** [comment]
>  GitGuardian has uncovered 1 secret following the scan of your pull request.

I marked this secret as a test credential.

**Pull request merged**